KCCM: fix transient node addition + removal while syncing load balancers

alexanderConstantinescu opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->


/kind bug

#### What this PR does / why we need it:

This PR adds one test case which exemplifies how a Node addition + Service update while an on-going Node sync is occurring can both add and remove the same node from the load balancer's set of backend nodes. 

The second commit fixes the issue by changing `lastSyncedNodes` to be keyed per service key. This can then be accessed from both the service and node control loops and synchronize the nodes used for any load balancer sync, between both controllers. This also fixes the broken test case added in commit 1. 

/sig network
/sig cloud-provider
/cc @aojea 
/assign @thockin   

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #121090

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
KCCM: fix transient node addition + removal caused by #121090 while syncing load balancers on large clusters with a lot of churn
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
None
```

alexanderConstantinescu committed:
KCCM: add test validating slow node sync issue

alexanderConstantinescu committed:
KCCM: fix slow node sync + service update

thockin commented on staging/src/k8s.io/cloud-provider/controllers/service/controller_test.go:
```diff
@@ -897,8 +916,22 @@ func compareUpdateCalls(t *testing.T, left, right []fakecloud.UpdateBalancerCall
 	}
 }
 
+func compareHostSets(t *testing.T, left, right []*v1.Node) bool {
```
I don't understand this - it claims to compare sets, but it returns true as long as any one of left matches one of right - am I misreading it?

thockin replied:
Indeed, if I force a test case where the last node differs it still passes.

alexanderConstantinescu replied:
Sorry, rectified this.

thockin commented on staging/src/k8s.io/cloud-provider/controllers/service/controller_test.go:
```diff
@@ -146,13 +147,13 @@ func defaultExternalService() *v1.Service {
-
-func newController() (*Controller, *fakecloud.Cloud, *fake.Clientset) {
-	cloud := &fakecloud.Cloud{}
+func newController(stopCh <-chan struct{}, objects ...runtime.Object) (*Controller, *fakecloud.Cloud, *fake.Clientset) {
```
comment on what objects is?

thockin commented on staging/src/k8s.io/cloud-provider/controllers/service/controller_test.go:
```diff
@@ -1329,8 +1367,95 @@ func TestNeedsCleanup(t *testing.T) {
+	expectedUpdateCalls := []fakecloud.UpdateBalancerCall{
+		{Service: service1, Hosts: []*v1.Node{node1, node2}},
+		{Service: service2, Hosts: []*v1.Node{node1, node2, node3}},
+		{Service: service2, Hosts: []*v1.Node{node1, node2, node3}},
```
Comment why this last is duplicated and why we don't see service1 with all 3 ?

aojea commented on staging/src/k8s.io/cloud-provider/fake/fake.go:
```diff
@@ -86,6 +86,7 @@ type Cloud struct {
 	ExternalIP    net.IP
 	Balancers     map[string]Balancer
 	UpdateCalls   []UpdateBalancerCall
+	UpdateCallCh  chan UpdateBalancerCall
```
this is not great for a general mock, also it will depend on the users to create the channel and buffered or it will block, can we find a better way than this?

alexanderConstantinescu replied:
I couldn't at least. We need a way to be able to trigger an update on the service while the node sync is processing, besides using a channel I didn't have any other ideas.

aojea replied:
no ideas sorry,

aojea commented on staging/src/k8s.io/cloud-provider/controllers/service/controller_test.go:
```diff
@@ -146,13 +147,13 @@ func defaultExternalService() *v1.Service {
-	cloud := &fakecloud.Cloud{}
+func newController(stopCh <-chan struct{}, objects ...runtime.Object) (*Controller, *fakecloud.Cloud, *fake.Clientset) {
+	cloud := &fakecloud.Cloud{
+		UpdateCallCh: make(chan fakecloud.UpdateBalancerCall, 1000),
```
this is a strong assumption for consumers of the mock

aojea commented on staging/src/k8s.io/cloud-provider/controllers/service/controller_test.go:
```diff
@@ -162,26 +163,36 @@ func newController() (*Controller, *fakecloud.Cloud, *fake.Clientset) {
+	}
+
+	informerFactory.Start(stopCh)
+	informerFactory.WaitForCacheSync(stopCh)
```
this is a noop if you don't pass the cache you want to wait to sync

https://github.com/alexanderConstantinescu/kubernetes/blob/4fb75e2f0d9a36c47edcf65f89bb92f20274ee56/staging/src/k8s.io/client-go/tools/cache/shared_informer.go#L211

but for unit tests we always have to mock it 
```
controller.nodeListerSynced = alwaysReady
	controller.serviceListerSynced = alwaysReady
```

alexanderConstantinescu replied:
Not sure I follow, this calls: https://github.com/alexanderConstantinescu/kubernetes/blob/4fb75e2f0d9a36c47edcf65f89bb92f20274ee56/staging/src/k8s.io/client-go/informers/factory.go#L141 

Which will just utilise `informer.HasSynced`, here: https://github.com/alexanderConstantinescu/kubernetes/blob/4fb75e2f0d9a36c47edcf65f89bb92f20274ee56/staging/src/k8s.io/client-go/informers/factory.go#L157

alexanderConstantinescu replied:
It's not calling that. What you are linking to is the function `WaitForCacheSync`, the code is calling the pointer receiver `WaitForCacheSync` - which is what I linked to here: https://github.com/alexanderConstantinescu/kubernetes/blob/4fb75e2f0d9a36c47edcf65f89bb92f20274ee56/staging/src/k8s.io/client-go/informers/factory.go#L141

aojea replied:
doh, I should stop checking these things at night, sorry for the noise

aojea commented on staging/src/k8s.io/cloud-provider/controllers/service/controller_test.go:
```diff
@@ -1329,8 +1358,95 @@ func TestNeedsCleanup(t *testing.T) {
+	cloudProvider.RequestDelay = 4 * duration
+
+	/*
+		This tests a service update while a node sync is happening. If we have multiple
+		services to process from a node sync: each service will experience a sync delta.
+		If a new Node is added and a service is synced while this happens: we want to
+		make sure that the slow node sync never removes the Node from LB set because it
+		has stale data.
+	*/
```
This test is using the fakeClientset that has a lot of known issues and is mocking most of the behaviors but is far from being real, I would not trust this test as a reproducer of what you want to achieve.

The best framework for reproducing these problems is the integration tests that exercises the real code paths  https://github.com/kubernetes/kubernetes/pull/120630#issuecomment-1729208508

alexanderConstantinescu replied:
>The best framework for reproducing these problems is the integration tests that exercises the real code paths https://github.com/kubernetes/kubernetes/pull/120630#issuecomment-1729208508 

Integration tests will have a hard time reproducing this problem. This problem occurs when we have a lot nodes and service LBs - so when the node sync takes a long time to process and we get a service update while this happens. This might require creating a test cluster with hundreds of nodes and LBs, and even then: it would depend on how quickly the cloud provider is able to process all LB operations. 

With a unit test we can actually mock the behavior and reproduce the edge case consistently. And logically I think we can all agree that the edge case is real and that the problem exists....I just don't know how to consistently reproduce that with an E2E/integration test.

alexanderConstantinescu commented on staging/src/k8s.io/cloud-provider/controllers/service/controller.go:
```diff
@@ -439,16 +439,33 @@ func (c *Controller) ensureLoadBalancer(ctx context.Context, service *v1.Service
 		c.eventRecorder.Event(service, v1.EventTypeWarning, "UnAvailableLoadBalancer", "There are no available nodes for LoadBalancer")
 	}
-
 	// - Only one protocol supported per service
```
ACK, will remove this comment then.

alexanderConstantinescu committed:
Store nodes before calling EnsureLoadBalancer

I am having difficulties convincing myself if this is better or worse.
I didn't implement this originally because I didn't want to store nodes that
we weren't sure we've configured. However: if EnsureLoadBalancer fails we
should retry the call from the service controller. Doing it like this might
save us one update call from the node controller side for calls which have
already started executing from the service controller's side...is this really
that expensive at this point though? Is it really that dangerous to not do
either, given that we retry failed calls? Ahhhhh!!! Opinions, please! Help, please!

thockin commented on staging/src/k8s.io/cloud-provider/controllers/service/controller_test.go:
```diff
@@ -901,8 +925,28 @@ func compareUpdateCalls(t *testing.T, left, right []fakecloud.UpdateBalancerCall
+	if len(left) != len(right) {
+		return false
+	}
+	for _, lHost := range left {
```
for followup?

staging/src/k8s.io/apimachinery/pkg/util/sets:

```
l := sets.New(left...)
r := sets.New(right...)
return reflect.DeepEqual(l, r)
```

?

alexanderConstantinescu replied:
OK, this comment made me go down a two hour rabbithole into the depths of client-go, and I still haven't come out sane. At first I thought: "makes total sense, let me update the code" - except the test started failing, at which point I started my journey. 

By doing this: `reflect.DeepEqual` will start comparing the pointers of each item in the `set`s (read: `map`s) - and these differ. That is to say: the objects/pointers of `node1`/`node2` as retrieved in the informer cache are not the objects/pointers as the test defines at the top, here: https://github.com/kubernetes/kubernetes/blob/b8c22aa0e2b565b98daafdf6f1b498c5e4385f48/staging/src/k8s.io/cloud-provider/controllers/service/controller_test.go#L1398-L1399

At one point (I don't know where...) something must be doing a `.DeepCopy` on these objects when populating the cache.

Given that the current state of the code does a `reflect.DeepEqual(nodeLeft, nodeRight)` the comparison changes and underlying values of the struct fields are compared (see: https://github.com/golang/go/blob/master/src/reflect/deepequal.go#L123-L127) - so the test passes. 

My question is: does this mean the test setup is wrong, or rather that this comment doesn't really work?

thockin reviewed (commented):
@aojea are you OK to followup the test questions?  The controller LGTM

aojea commented:
> @aojea are you OK to followup the test questions? The controller LGTM

yes, let's see if we can make the mock addition optional at least

aojea commented on staging/src/k8s.io/cloud-provider/controllers/service/controller.go:
```diff
@@ -794,20 +811,12 @@ func nodesSufficientlyEqual(oldNodes, newNodes []*v1.Node) bool {
 
 	doWork := func(piece int) {
-		if shouldRetry := c.nodeSyncService(services[piece], c.lastSyncedNodes, nodes); !shouldRetry {
+		if shouldRetry := c.nodeSyncService(services[piece]); !shouldRetry {
```
this seems like an important penalty, no?

now we are going to list and filter all nodes per each number of Services

thockin replied:
but only from cache, right?

alexanderConstantinescu replied:
Yes, only from the cache.

aojea replied:
👍

aojea commented on staging/src/k8s.io/cloud-provider/controllers/service/controller.go:
```diff
@@ -737,19 +744,29 @@ func (c *Controller) syncNodes(ctx context.Context, workers int) sets.String {
 	newNodes = filterWithPredicates(newNodes, getNodePredicatesForService(svc)...)
-	oldNodes = filterWithPredicates(oldNodes, getNodePredicatesForService(svc)...)
-
+	oldNodes := filterWithPredicates(c.getLastSyncedNodes(svc), getNodePredicatesForService(svc)...)
+	// Store last synced nodes without actually determining if we successfully
+	// synced them or not. Failed node syncs are passed off to retries in the
+	// service queue, so no need to wait. If we don't store it now, we risk
+	// re-syncing all LBs twice, one from another sync in the node sync and
+	// from the service sync
+	c.storeLastSyncedNodes(svc, newNodes)
```
these storeLastSyncedNodes and getLastSyncedNodes hold a lock now, so we can't really parallelize this nodeSyncService now

aojea replied:
otherwise a Service change will call `EnsureLoadBalancer` and a Node change will call `UpdateLoadBalancer` at the same time

aojea replied:
> That is allowed and shouldn't be a problem as long as the nodes passed are always up-to-date and synchronized between both calls. The issue this PR tries to solve is: synchronizing the nodes which are passed.

sorry for keep asking questions but this controller is always super confusing to me, is that a safe assumption " shouldn't be a problem as long as the nodes passed are always up-to-date and synchronized between both calls"

alexanderConstantinescu replied:
> is that a safe assumption

I think so. We now keep a cache of nodes for each service, which is synchronized and protected by a lock. Given my overall analysis and understanding of this controller: I don't see how the nodes passed to these API calls can now differ, and moreover: how we can end up removing a node which needs to be in the load balancer set. Famous last words, I know.... :)

aojea commented:
Let's do a step back, I think that something is odd here, 
1.  This controller react on Nodes and Services
2. The Service logic seems relatively easy
https://github.com/kubernetes/kubernetes/blob/16623991318567cd6c29e2114867e7654f25c825/staging/src/k8s.io/cloud-provider/controllers/service/controller.go#L873-L876

3. Nodes are more complex, the worker process nodes and requeue some Services https://github.com/kubernetes/kubernetes/blob/16623991318567cd6c29e2114867e7654f25c825/staging/src/k8s.io/cloud-provider/controllers/service/controller.go#L272-L273

syncNodes reconciles the whole cluster

https://github.com/kubernetes/kubernetes/blob/16623991318567cd6c29e2114867e7654f25c825/staging/src/k8s.io/cloud-provider/controllers/service/controller.go#L716-L729

and do that by calling to `updateLoadbalacerHosts` https://github.com/kubernetes/kubernetes/blob/16623991318567cd6c29e2114867e7654f25c825/staging/src/k8s.io/cloud-provider/controllers/service/controller.go#L791-L794

that parallelize `nodeSyncService` for all services

https://github.com/kubernetes/kubernetes/blob/16623991318567cd6c29e2114867e7654f25c825/staging/src/k8s.io/cloud-provider/controllers/service/controller.go#L809-L812

that interesting seems to call without actually `lockedUpdateLoadBalancerHosts` holding the lock

https://github.com/kubernetes/kubernetes/blob/16623991318567cd6c29e2114867e7654f25c825/staging/src/k8s.io/cloud-provider/controllers/service/controller.go#L749-L754

and this updates the loadbalancer

https://github.com/kubernetes/kubernetes/blob/16623991318567cd6c29e2114867e7654f25c825/staging/src/k8s.io/cloud-provider/controllers/service/controller.go#L826-L838

So it seems we have to paths updating loadbalancers, so it will always race, if I create a Service and I update a Node at the same time, both events will follow different paths, no?

If my reasoning is correct the problem of races is not solvable meanwhile there are two independent paths in the controller updating loadbalancer, why don't we have just one place to sync the loadbalancer and the controller on the Node worker checks the impacted Service and just requeue the Service

alexanderConstantinescu commented on staging/src/k8s.io/cloud-provider/controllers/service/controller.go:
```diff
@@ -93,10 +93,10 @@ type Controller struct {
-	lastSyncedNodes []*v1.Node
+	// the last synced set of nodes per service key. This is accessed from the
+	// service and node controllers, hence it is protected by a lock.
+	lastSyncedNodes     map[string][]*v1.Node
```
We can address this in a follow up of course, but IMO this is fine. The only services used as keys in this map are `type: LoadBalancer` - so it could be considered "a lot more information" on clusters which have a lot of services like that, but usually that's not the case.

alexanderConstantinescu commented:
> If my reasoning is correct the problem of races is not solvable meanwhile there are two independent paths in the controller updating loadbalancer, why don't we have just one place to sync the loadbalancer and the controller on the Node worker checks the impacted Service and just requeue the Service

I _think_ the reason they have seperate paths is because we don't want to block reconciling service updates, because we are stuck processing all services/load balancers from a node update. That would might have any changes on the service object (ex adding an annotation) take minutes to process if the controllers are busy processing _any_ node update.

Moreover: my goal here was to fix the problem at hand (possibly also back-port this), so I'd like the implementation to remain as small as possible and avoid refactoring.

aojea commented:
> _think_ the reason they have seperate paths is because we don't want to block reconciling service updates, because we are stuck processing all services/load balancers from a node update. That might have any changes on the service object (ex adding an annotation) take minutes to process if the controllers are busy processing _any_ node update.

but why do we want to fix a corner case if the fundamental logic is wrong https://github.com/kubernetes/kubernetes/pull/121091/files#r1373296928

alexanderConstantinescu commented:
> but why do we want to fix a corner case if the fundamental logic is wrong

I don't know that the logic is wrong. Any node update _needs_ to re-sync all load balancers (so will take a lot of time). Any service update only needs to sync one load balancer. We definitely don't want any service update to wait on the node update to finish cycling through all load balancers before starting, because that will slow down that service update by quite a lot. 

Another idea I had was using a`map[string]sync.Mutex`, where the key is the service key. In that case we could lock the API call, because it would only be locked for that particular service. But the core issue here is really just synchronizing the nodes and making sure that they are always correct and up-to-date when performing the API call

thockin reviewed (commented):
I also get confused in here, so I took some notes:

Before this PR:

Service update:
```
c.processNextServiceItem()
	c.syncService(key)
		c.processServiceCreateOrUpdate(svc, key)
			c.syncLoadBalancerIfNeeded(svc, key)
				c.ensureLoadBalancer(svc)
					nodes= listWithPredicates(c.nodeLister, getNodePredicatesForService(service)...)
					c.balancer.EnsureLoadBalancer(svc, nodes)
```

Nodeupdate:
```
processNextNodeItem()
	c.syncNodes()
		svcs = c.cache.allServices()
		c.updateLoadBalancerHosts(svcs)
			nodes = listWithPredicates(c.nodeLister)
			workqueue.Parallelize c.nodeSyncService(svcs[i], c.lastSyncedNodes, nodes)
				newNodes = filterWithPredicates(newNodes, getNodePredicatesForService(svc)...)
				c.lockedUpdateLoadBalancerHosts(svc, newNodes)
					c.balancer.UpdateLoadBalancer(svc, nodes)
```

I don't know why one path calls EnsureLoadBalancer and one calls UpdateLoadBalancer.  They both handle errors differently, for sure.  It feels like they should converge at `c.ensureLoadBalancer(svc)`, right?

After this PR:

Nodeupdate:
```
processNextNodeItem()
	c.syncNodes()
		svcs = c.cache.allServices()
		c.updateLoadBalancerHosts(svcs)
			workqueue.Parallelize c.nodeSyncService(svcs[i])
				newNodes = listWithPredicates(c.nodeLister)
				newNodes = filterWithPredicates(newNodes, getNodePredicatesForService(svc)...)
				c.lockedUpdateLoadBalancerHosts(svc, newNodes)
					c.balancer.UpdateLoadBalancer(svc, nodes)
```

I don't think this is worse.  It still leaves a lot of ??? in there, though.

aojea commented:
> I don't think this is worse. It still leaves a lot of ??? in there, though.

keep digging and see how you get still more ???? , I only checked the gce cloud provider code and both `UpdateLoadBalancer` and `EnsureLoadBalancer` hold a shared lock for the internal loadbalancers implementation for both Update and Ensure and  only is locked on Ensure for the external loadbalancers

I don't have doubts this is better, I just don't think is ok having two reconcile loops stomping into each other in this very expensive and slow API calls (at least in GCE forwarding , rule bakends service, ...) can be O(minutes) ... we had issues with customers with a considerable number of loadbalancers complaining about the time it takes to sync the loadbalancers

aojea commented:
Current mode of having two loops, one for node and one for service is causing issues downstream on customers and is super complex to troubleshoot, since the implementations details of each cloud provider is an important factor since some operations insde the loadbalancer methods lock, per example in GCE,  causing all the operations to be serialized and since there are 5 workers , in large cluster there can be changes that take more than one hour

If we remove the node loop this is no longer needed, I rather prefer to invest on fixing the root problem

alexanderConstantinescu commented:
> Actually this should be better in that you are less likely to end up with a stale set of backends, thus needing toi run again - right?

Right, that's the intention. 

> If we remove the node loop this is no longer needed, I rather prefer to invest on fixing the root problem

I don't really follow what the proposal is, are you saying we should stop watching nodes?

thockin commented:
I spent some time today talking to Antonio, and I think I understand his point.

This code today is non-obvious and (see my notes above) seems ripe for "fixing".  But the "fixes" are very hard to reason about without also looking into the per-cloud code.  I realized why one path ends up at "ensure" and another at "update" - because in the case of a node change, we know the service existed AND had a load-balancer, so we don't need to call "ensure".  If we "fix" this, it *might* cause 3x or 4x the number of cloud-provider API calls doing work taht doesn't need to be done.  That will pass e2e but will explode at scale, causing an outage for folks like the GKE, EKS, AKS, etc teams.

Or it might be fine.

So when he says "fixing the root problem" I hear "how would we design this if we were starting over?"  Maybe we would have a single channel, into which we feed service names which need LB updates, and let the cloud-provider code figure out how to parallelize and when to call ensure vs. update.?  But that would be an incompatible change, requiring providers to opt-in.

As to this PR, I don't think it makes that better or worse, but as we "simplify" some parts of this, it makes us question the rest.

@aojea - did I capture it?  Are you OK to proceed with this PR, and we can consider how to fix this properly in a different context?  Or am I still missing it?

thockin commented:
I'm going to approve, and let @aojea LGTM or not

/approve

thockin commented:
Test fail seems real?

/retest

aojea commented:
/lgtm

@thockin you summarized it perfectly, let's fix the problem that we know then

aojea commented:
/retest

alexanderConstantinescu commented:
> Test fail seems real?

Yes, I am debugging. It seems to point to issue with nodes:

```
I1031 11:33:16.358642   94432 event.go:376] "Event occurred" object="test-service-update-load-balancer-ip-mode/test-update-load-balancer-ip-mode" fieldPath="" kind="Service" apiVersion="v1" type="Normal" reason="EnsuringLoadBalancer" message="Ensuring load balancer"
I1031 11:33:16.362486   94432 event.go:376] "Event occurred" object="test-service-update-load-balancer-ip-mode/test-update-load-balancer-ip-mode" fieldPath="" kind="Service" apiVersion="v1" type="Warning" reason="UnAvailableLoadBalancer" message="There are no available nodes for LoadBalancer"
    loadbalancer_test.go:526: unexpected load balancer status
```

alexanderConstantinescu commented:
/test pull-kubernetes-unit

alexanderConstantinescu commented:
I worked out the problem. The load balancer integration tests didn't initialize a buffered `UpdateCallCh` so it was blocking the update calls. To resolve this I added:  https://github.com/kubernetes/kubernetes/blob/e0da17626e0e9b7b0c5d578fcfbc2270bd06a965/staging/src/k8s.io/cloud-provider/fake/fake.go#L232-L234
I figured this won't block people who write tests which depend on the fake cloud provider and forget to initialize a buffered `UpdateCallCh`.

aojea commented on staging/src/k8s.io/cloud-provider/fake/fake.go:
```diff
@@ -222,13 +224,21 @@ func (f *Cloud) EnsureLoadBalancer(ctx context.Context, clusterName string, serv
+	defer f.Lock.Unlock()
+	update := UpdateBalancerCall{service, nodes}
+	f.UpdateCalls = append(f.UpdateCalls, update)
+	if cap(f.UpdateCallCh) > 0 {
+		f.UpdateCallCh <- update
+	}
```
this is super hacky :/

aojea replied:
I already expressed some concerns in https://github.com/kubernetes/kubernetes/pull/121091/files#r1355907804 and this makes me have more doubts we should merge this and modify the mock for every users just for one test,

Can we add a hook so this specific test injects its own logic, so we don't make this part of any other test?

alexanderConstantinescu replied:
> this is super hacky :/

Yeah, that wasn't great I agree. 

I've updated the PR with the possibility for callers to define the callback. The only test currently defining the callback is the "slow test" I've added in this PR. We should not be impacting other tests now.

aojea commented on staging/src/k8s.io/cloud-provider/fake/fake.go:
```diff
@@ -86,6 +86,7 @@ type Cloud struct {
 	ExternalIP    net.IP
 	Balancers     map[string]Balancer
 	UpdateCalls   []UpdateBalancerCall
+	UpdateCallCb  func(UpdateBalancerCall)
```
I think this is much better
/lgtm

aojea replied:
/hold

aojea commented on staging/src/k8s.io/cloud-provider/fake/fake.go:
```diff
@@ -200,6 +201,7 @@ func (f *Cloud) GetLoadBalancerName(ctx context.Context, clusterName string, ser
 // EnsureLoadBalancer is a test-spy implementation of LoadBalancer.EnsureLoadBalancer.
 // It adds an entry "create" into the internal method call record.
 func (f *Cloud) EnsureLoadBalancer(ctx context.Context, clusterName string, service *v1.Service, nodes []*v1.Node) (*v1.LoadBalancerStatus, error) {
+	f.markUpdateCall(service, nodes)
```
this was not called before

alexanderConstantinescu replied:
I added this because the service sync path ends up calling `EnsureLoadBalancer` and I need the call to get registered so that I can validate this in the tests

aojea replied:
yeah, but we can not modify a general mock to fit into our custom logic in the test, if you want to add custom logic in the Ensure method just add an Ensure callback so people can override it

alexanderConstantinescu replied:
> yeah, but we can not modify a general mock to fit into our custom logic in the test

At some point someone modified the general mock to fit the custom logic of these tests...why can't I not do that now?

aojea replied:
> At some point someone modified the general mock to fit the custom logic of these tests.

the problem is not customize the mock , the problem is to add a the updateCallback on the Ensure method, just add an enureCallback so it is correct  ... test code is the same as production code

alexanderConstantinescu replied:
OK, I've updated the PR and added explicit field for the ensure calls (so both `EnsureCalls`/ `EnsureCallCb`)

aojea commented:
/retest

This is a bugfix so we can merge after code freeze, I'm just holding because I have one question in the test implementation,  but the code is good to go

aojea commented on staging/src/k8s.io/cloud-provider/fake/fake.go:
```diff
@@ -222,13 +224,21 @@ func (f *Cloud) EnsureLoadBalancer(ctx context.Context, clusterName string, serv
 // UpdateLoadBalancer is a test-spy implementation of LoadBalancer.UpdateLoadBalancer.
 // It adds an entry "update" into the internal method call record.
 func (f *Cloud) UpdateLoadBalancer(ctx context.Context, clusterName string, service *v1.Service, nodes []*v1.Node) error {
+	f.markUpdateCall(service, nodes)
```
this was running after `addCall`

alexanderConstantinescu replied:
Yes, but if someone specifies a `RequestDelay` what we want is to register the call and then wait `RequestDelay`. The existing tests never utilised `RequestDelay`, so it didn't matter to them that we registered the call after having waited - but for the test case I've added, it does.

aojea replied:
>  he existing tests never utilised RequestDelay, so it didn't matter to them that we registered

but is not about the existing tests, future tests may want to use it and it will not work, this is not correct to modify a behavior like this. Since our test does not use RequestDelay it does not matter for us the order, so we should keep it

alexanderConstantinescu replied:
>  future tests may want to use it and it will not work

What will not work? 

> Since our test does not use RequestDelay it does not matter for us the order, so we should keep it

I don't understand: what is "our test"?

aojea replied:
> What will not work?

this change runs `markUpdateCall` before `RequestDelay` , this test does not set `RequestDelay`, why do you have to change the order? am I missing something?

alexanderConstantinescu replied:
The test I've added (`TestSlowNodeSync`) does set `RequestDelay`. The reason `markUpdateCall` needs to run before waiting `RequestDelay`, is because we need to mock the behaviour whereby we trigger a service sync while the node sync is processing. Essentially the test does the following:

1. Process a node sync. 
2. [node sync] `UpdateLoadBalancer` will be called for service1 (takes `RequestDelay` to process)
3. node3 is created 
4. service sync is triggered for service2
5. [service sync] `EnsureLoadBalancer` will be called for service2
6. [node sync] `UpdateLoadBalancer` will be called for service2 

If we run `markUpdateCall` after waiting `RequestDelay` we can't consistently mock the behaviour and might have the following instead:

1. We process a node sync. 
2. [node sync] `UpdateLoadBalancer` will be called for service1 
3. [node sync] `UpdateLoadBalancer` will be called for service2 
4. node3 is created 
5. service sync is triggered for service2
6. [service sync] `EnsureLoadBalancer` will be called for service2 

i.e: we need to create node3 and trigger the service sync for service2 _while_ the we are processing service1 from the node sync. This validates that we don't remove nodes from the service2 LB set, because the node sync has stale node data. This is the bug this PR is trying to fix.    

For all existing tests, this behaviour is never tested, so it doesn't matter if `markUpdateCall` is called before waiting `RequestDelay` or after.

aojea replied:
can't you add an updateCall that adds the delay instead of having to modify the behavior?

alexanderConstantinescu replied:
Done

alexanderConstantinescu commented:
/test pull-kubernetes-unit

@aojea : can you please have a look at this PR again, I've addressed your remaining comments. Thanks in advance!

aojea commented on staging/src/k8s.io/cloud-provider/controllers/service/controller_test.go:
```diff
@@ -1333,8 +1380,113 @@ func TestNeedsCleanup(t *testing.T) {
+	defer close(stopCh)
+	defer close(updateCallCh)
+
+	duration := time.Millisecond
```
this is very small and use to flake on CI with load , did you run the test with `stress` to verify it does not flake? see instructions in https://gist.github.com/liggitt/6a3a2217fa5f846b52519acfc0ffece0#running-unit-tests-to-reproduce-flakes

alexanderConstantinescu replied:
I've removed the sleet completely and used a channel instead to coordinate the syncs between both controllers.

alexanderConstantinescu committed:
Refine test case

alexanderConstantinescu commented:
@aojea : I've updated the test cases. These are now the results:

```
 % go test -timeout 30s -run ^TestSlowNodeSync$ k8s.io/cloud-provider/controllers/service -count=100 -race
ok  	k8s.io/cloud-provider/controllers/service	12.166s
% stress ./service.test -test.run TestSlowNodeSync
5s: 34 runs so far, 0 failures
10s: 99 runs so far, 0 failures
15s: 164 runs so far, 0 failures
20s: 228 runs so far, 0 failures
25s: 292 runs so far, 0 failures
30s: 372 runs so far, 0 failures
35s: 436 runs so far, 0 failures
40s: 500 runs so far, 0 failures
45s: 564 runs so far, 0 failures
50s: 635 runs so far, 0 failures
55s: 708 runs so far, 0 failures
1m0s: 772 runs so far, 0 failures
1m5s: 836 runs so far, 0 failures
1m10s: 903 runs so far, 0 failures
1m15s: 971 runs so far, 0 failures
1m20s: 1035 runs so far, 0 failures
1m25s: 1108 runs so far, 0 failures
1m30s: 1175 runs so far, 0 failures
1m35s: 1243 runs so far, 0 failures
1m40s: 1307 runs so far, 0 failures
1m45s: 1371 runs so far, 0 failures
1m50s: 1447 runs so far, 0 failures
1m55s: 1515 runs so far, 0 failures
2m0s: 1579 runs so far, 0 failures
2m5s: 1643 runs so far, 0 failures
2m10s: 1707 runs so far, 0 failures
2m15s: 1772 runs so far, 0 failures
2m20s: 1836 runs so far, 0 failures
2m25s: 1915 runs so far, 0 failures
2m30s: 1979 runs so far, 0 failures
2m35s: 2044 runs so far, 0 failures
2m40s: 2108 runs so far, 0 failures
2m45s: 2174 runs so far, 0 failures
2m50s: 2240 runs so far, 0 failures
2m55s: 2316 runs so far, 0 failures
3m0s: 2380 runs so far, 0 failures
3m5s: 2446 runs so far, 0 failures
3m10s: 2514 runs so far, 0 failures
3m15s: 2578 runs so far, 0 failures
3m20s: 2642 runs so far, 0 failures
3m25s: 2718 runs so far, 0 failures
3m30s: 2786 runs so far, 0 failures
3m35s: 2850 runs so far, 0 failures
3m40s: 2914 runs so far, 0 failures
3m45s: 2978 runs so far, 0 failures
3m50s: 3042 runs so far, 0 failures
3m55s: 3106 runs so far, 0 failures
4m0s: 3170 runs so far, 0 failures
4m5s: 3234 runs so far, 0 failures
4m10s: 3314 runs so far, 0 failures
4m15s: 3378 runs so far, 0 failures
4m20s: 3442 runs so far, 0 failures
4m25s: 3506 runs so far, 0 failures
4m30s: 3570 runs so far, 0 failures
4m35s: 3634 runs so far, 0 failures
4m40s: 3698 runs so far, 0 failures
4m45s: 3762 runs so far, 0 failures
4m50s: 3826 runs so far, 0 failures
4m55s: 3890 runs so far, 0 failures
5m0s: 3970 runs so far, 0 failures
5m5s: 4034 runs so far, 0 failures
5m10s: 4098 runs so far, 0 failures
5m15s: 4162 runs so far, 0 failures
^C
```

aojea commented on staging/src/k8s.io/cloud-provider/controllers/service/controller_test.go:
```diff
@@ -1333,8 +1380,96 @@ func TestNeedsCleanup(t *testing.T) {
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		controller.syncNodes(context.TODO(), 1)
```
this calls the UpdateCallCb and sens the `syncService <- impactedService` through the channel

aojea commented on staging/src/k8s.io/cloud-provider/controllers/service/controller_test.go:
```diff
@@ -1333,8 +1380,96 @@ func TestNeedsCleanup(t *testing.T) {
+		controller.syncNodes(context.TODO(), 1)
+	}()
+
+	key := <-syncService
```
this takes the service updated as signal to create a new node, the update is still being processed

aojea commented on staging/src/k8s.io/cloud-provider/controllers/service/controller_test.go:
```diff
@@ -1333,8 +1380,96 @@ func TestNeedsCleanup(t *testing.T) {
+	}
+
+	// Sync the service
+	if err := controller.syncService(context.TODO(), key); err != nil {
+		t.Fatalf("unexpected service sync error, err: %v", err)
+	}
+
+	wg.Wait()
```
we process the service now in parallel and wait until the `syncNodes` finish

aojea replied:
`syncNodes` will be blocked until `syncService`finish due to the `EnsureCallCb`

aojea commented:
/lgtm
/approve
/milestone v1.29

This qualifies for the milestone, it was approved before code freeze and blocked on test review that I unfortunatly delayed.
It is also a bugfix

alexanderConstantinescu commented:
/hold cancel
